File: code/ch09/9.2.2.4.long_substr_without_repeat_char.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
            int res = 0, left = 0, right = 0;
            int charset[256] = {-1};

            while (right < s.size() && left + res < s.size()) {
                left = max(left, charset[s[right]]);
                res  = max(res, right - left + 1);
                charset[s[right]] = right + 1;
                right++;
            }

            return res;
        }
};

File: code/ch09/9.2.2.4.long_substr_without_repeat_char.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int res = 0, left = 0, right = 0;
        Map<Character, Integer> map = new HashMap<>();
        while (right < s.length() && left < s.length() - res) {
            if (map.containsKey(s.charAt(right)))
                left = Math.max(left, map.get(s.charAt(right)) + 1);

            map.put(s.charAt(right), right);
            res = Math.max(res, right - left + 1);
            right++;
        }

        return res;
    }
}